applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py

```diff
@@ -198,31 +198,43 @@ def MarkKuttaElements(self):
             # Marking the elements under the trailing edge as kutta
             if(distance_to_wake < 0.0):
                 elem.SetValue(CPFApp.KUTTA, True)
+    @staticmethod
+    def CheckIfElemIsCutByWake(elem):
+        npos=0
+        nneg=0
+        distances = elem.GetValue(KratosMultiphysics.ELEMENTAL_DISTANCES)
+        for nodal_distance in distances:
+            if nodal_distance>0:
+                npos += 1
+            else:
+                nneg += 1
 
-    def ComputeMaximumX(self):
-        # This function computes the maximum x coordinate within
-        # all the trailing edge elements' centers
-        maximum_x_coordinate = -1e30
+        return nneg==1
+
+    def FindTrailingEdgeElementId(self):
+        # This function finds the trailing edge element that is cut by the wake
 
         # Loop over the elements touching the trailing edge
         for elem in self.trailing_edge_model_part.Elements:
-            # Find the element touching the trailing edge with maximum x coordinate
-            if(elem.GetGeometry().Center().X > maximum_x_coordinate):
-                maximum_x_coordinate = elem.GetGeometry().Center().X
+            # Find the element touching the trailing edge that is cut by the wake
+            if (elem.GetValue(KratosMultiphysics.CompressiblePotentialFlowApplication.WAKE)):
+                if(self.CheckIfElemIsCutByWake(elem)):
+                    te_elem_id=elem.Id
 
-        return maximum_x_coordinate
+        return te_elem_id
 
     def MarkWakeTEElement(self):
         # This function finds the trailing edge element that is further downstream
-        # and marks it as wake trailing edge element
+        # and marks it as wake trailing edge element. The rest of trailing edge elements are
+        # unassigned from the wake.
 
-        # Compute the maximum x coordinate within all the trailing edge
-        # elements' centers
-        maximum_x_coordinate = self.ComputeMaximumX()
+        # Find element in the trailing edge that is completely cut by the wake
+        TrailingEdgeElementID = self.FindTrailingEdgeElementId()
 
         for elem in self.trailing_edge_model_part.Elements:
-            if(elem.GetGeometry().Center().X < maximum_x_coordinate):
-                elem.SetValue(CPFApp.WAKE, False)
-            else:
-                elem.Set(KratosMultiphysics.STRUCTURE)
-                elem.SetValue(CPFApp.KUTTA, False)
+            if (elem.GetValue(CPFApp.WAKE)):
+                if(elem.Id==TrailingEdgeElementID): #TE Element
```

Comment:
I think you do not need the function FindTrailingEdgeElementId() at all, you can call CheckIfElemIsCutByWake directly here.

```suggestion
                if(self.CheckIfElemIsCutByWake(elem)): #TE Element
```

Reply:
@marcnunezc check if this makes sense to you, I might be wrong! I tried it locally in my machine and the test was still running.